Mass-edit road selection panel

kianzarrin opened the pull request:
fixes #542

~~I was unable to fully fix mass edit overlay so I disabled it and created issue #789~~ 
I fixed all other problems.
This is ready for review.

Added Identical road selection UI to `roadAdjust` panel and `roadWorldInfoPanel`. See https://github.com/krzychu124/Cities-Skylines-Traffic-Manager-President-Edition/issues/542#issuecomment-578475576 for pictures.

### expected behavior
- Both panels trigger tutorial message (same).
- ~~activation of the panel in `roadAdjust` activates mass edit overlay so that user can see relevant traffic rules. same is true when user clicks a button. note that roadAdjust is in turn automatically activated when user tries to modify road selection.~~
- clicking a button will apply traffic rules and puts it in active state. clicking it again deactivates it and clears relevant traffic rules. Clear does not have undo mode.

### Scope:
Since there is so much new user interface here there is a lot of potential for new feature [EDIT: see #691]. Therefore I have to severely limit the scope of this review by limiting some of the functionality. The objective is not to implement all missing features in TMPE! Its to provide the base UI interface and lay out the foundation for the rest of the work.
~~EDIT: speaking of scope priority roads that turn on junctions are not supported. Roundabout paths on the other hand can go through different routes.~~
EDIT: paths that turn left/right on junctions are supported.


### further changes
- Road selection util is an interface that exposes the list of selected segments. Also it provides an event that is invoked whenever selection changes. EDIT: in future we may need to promote to manger in order to serialize intents see  #691
- Linear sprite button now has an **optional** disabled state.
- ~~implemented simplistic undo functionality for roundabout. (I worked around  #623)~~ EDIT:  with #623 resolved, I am able to clear traffic rules (its more basic than restore/undo see #568) 
- I moved the code to put priority signs responding to shift-click from PrioirtySingsTool.cs into PrioirtyRoad.cs . I modified it so that it I can use it by my road selection panel.

### Translation keys:
These keys have been added in another PR.:

```
Menu.csv
"RoadSelection.Tooltip:Clear"=>"Clear traffic rules related to high priority road" // for now!
"RoadSelection.Tooltip:Stop entry"=>"applies stop signs on connected roads"
"RoadSelection.Tooltip:Yield entry"=>"applies Yield signs on connected roads"
"RoadSelection.Tooltip:High priority"=>"Applies traffic rules for high priority road(customizable in options)"
"RoadSelection.Tooltip:Roundabout"=>"applies all traffic rules for roundabout. button is disabled when selection is not a roundabout."

tutorial.csv
"TMPE_TUTORIAL_BODY_RoadSelection"=>"   Clicking each of the buttons(except the clear button) applies traffic rules. clicking the same button again clears those traffic rules. Each button is enabled only when there is a valid selection of road/roundabout.\n\n   To select a road, click on the green road selection button."
"TMPE_TUTORIAL_HEAD_RoadSelection"=>"TM:PE Mass-edit Road Selection Panel"
```

kianzarrin commented on TLM/TLM/LoadingExtension.cs:
```diff
@@ -333,21 +333,21 @@ public class ManualHarmonyPatch {
-                if (removeVehicleButtonExtender != null) {
-                    Object.Destroy(removeVehicleButtonExtender, 10f);
+                var gameObject = UIView.GetAView().gameObject;
+                void Destroy<T>()
```
I like ugly repeated code ... NOT!

kianzarrin commented on TLM/TLM/UI/SubTools/PrioritySignsTool.cs:
```diff
@@ -51,95 +51,26 @@ public PrioritySignsTool(TrafficManagerTool mainTool)
+                //    return;
             }
             if (shiftDown) {
-                var primaryPrioType = PriorityType.None;
```
moved this code into a method so that I can call it from my road selection panel.

krzychu124 commented on TLM/TLM/UI/RoadSelectionPanels.cs:
```diff
@@ -0,0 +1,313 @@
+
+        private void HidePriorityRoadToggle(UIComponent priorityRoadToggle) {
+            priorityRoadToggle.eventVisibilityChanged +=
+                (component, value) => component.isVisible = false;
```
You are creating memory leak here which will prevent from destroying panel.

krzychu124 commented on TLM/TLM/UI/RoadSelectionPanels.cs:
```diff
@@ -0,0 +1,313 @@
+                        ToolsModifierControl.toolController.CurrentTool = previousTool;
+                        previousTool = null;
+                    }
+                };
```
Again, memory leak.

krzychu124 commented on TLM/TLM/UI/RoadSelectionPanels.cs:
```diff
@@ -0,0 +1,313 @@
+                    if (value) {
+                        TrafficManagerTool.ShowAdvisor("RoadSelection");
+                    }
+                };
```
Memory leak...

aubergine10 commented:
@krzychu124 I think this PR should wait for v11.1 as it's going to need extensive testing (once v11 STABLE is published, we can start testing this on v11 LABS)

krzychu124 commented:
I created lang strings on `Crowdin` but set hidden since translation is not final yet (`PR` moved to `11.2`).

aubergine10 commented on TLM/TLM/TLM.csproj:
```diff
@@ -566,19 +566,20 @@
 xcopy /y "$(TargetDir)TMPE.RedirectionFramework.dll" "%25DEPLOYDIR%25"
 xcopy /y "$(TargetDir)CSUtil.Commons.dll" "%25DEPLOYDIR%25"
 xcopy /y "$(TargetDir)0TMPE.Harmony.dll" "%25DEPLOYDIR%25"
+
```
Is this to fix the double-load hotswap issue?

kianzarrin replied:
Yes. Hot swap still crashes the game though. I didn't mean to push in these changes. I would add comments.

kvakvs commented on TLM/TLM/Util/PriorityRoad.cs:
```diff
@@ -9,18 +9,86 @@ namespace TrafficManager.Util {
+            foreach(ushort segId in segmentList) {
+                foreach(bool bStartNode in Constants.ALL_BOOL) {
+                    ApplyPrioritySigns(segId, bStartNode);
+                }
```
I believe calling it twice with false and true will look simpler than a loop.
Up to you to judge, but simpler is better.

kianzarrin replied:
you are right. this is remnants of old code.

kvakvs commented on TLM/TLM/Util/PriorityRoad.cs:
```diff
@@ -51,11 +119,51 @@ public static class PriorityRoad {
             foreach (bool startNode in Constants.ALL_BOOL) {
                 ushort nodeId = netService.GetSegmentNodeId(segmentId, startNode);
-                FixJunction(nodeId);
+                FixHighPrioirtyJunction(nodeId);
```
Typo `prio-ir-ty`

kvakvs commented on TLM/TLM/Util/PriorityRoad.cs:
```diff
@@ -51,11 +119,51 @@ public static class PriorityRoad {
+        /// <param name="segmentId"></param>
+        /// <param name="otherSegmentId"></param>
+        /// <returns></returns>
+        private static ushort GetUnsharedNode(ushort segmentId, ushort otherSegmentId) {
```
Naming question. `Unshared` means that it was shared and now is not shared anymore. I.e. something has been done on that shared node. Maybe `GetSharedOrTheOtherNode`?

kvakvs commented on TLM/TLM/Util/PriorityRoad.cs:
```diff
@@ -51,11 +119,51 @@ public static class PriorityRoad {
+            }
+
+            foreach (ushort segmentId in segmentList) {
+                foreach (bool startNode in Constants.ALL_BOOL) {
```
Here foreach in ALL_BOOL is good because it groups multiple lines of code.

kvakvs commented on TLM/TLM/Util/PriorityRoad.cs:
```diff
@@ -141,6 +249,40 @@ public static class PriorityRoad {
+        ///  - semi-roundabout
+        ///  - split avenue into 2 oneway roads (only applicable to first or last node).
+        /// </summary>
+        public static void FixHighPrioirtyJunction(ushort nodeId, List<ushort> segmentList) {
```
naming typo: Prio-ir-ty

kvakvs commented on TLM/TLM/Util/PriorityRoad.cs:
```diff
@@ -141,6 +249,40 @@ public static class PriorityRoad {
+            }
+
+            if (nodeSegments.Count < 3) {
+                Log._Debug("FixJunction: This is not a junction.");
```
Who will read this debug message? User will never see it, developer will not understand it out of context. 
Remove before merge or convert into a proper error with code location.

kianzarrin replied:
Lets say you are trying to debug why HighPrio-ir-ity (lol) quick setup does not work. if you see in debug log "FixJunction: This is not a junction." it serves as very useful information. if you don't see this message it is also useful.
I can add nodeID.

kvakvs commented on TLM/TLM/Util/PriorityRoad.cs:
```diff
@@ -149,65 +291,74 @@ public static class PriorityRoad {
 
-            if (segmentList.Count < 3) {
+            if (nodeSegments.Count < 3) {
                 Log._Debug("FixJunction: This is not a junction.");
```
Same note. Who will be reading this message, and how to figure out its location? It is quite short, and reading log files to support some user will tell us nothing.

kvakvs commented on TLM/TLM/Util/PriorityRoad.cs:
```diff
@@ -149,65 +291,74 @@ public static class PriorityRoad {
+        /// <param name="nodeId">Junction to apply rules</param>
+        /// <param name="nodeSegments">list of segments. The first two elements are main/roundabout,
+        /// all other semenets are minor</param>
+        private static void FixHighPrioirtyJunctionHelper(ushort nodeId, List<ushort> nodeSegments) {
```
prio-ir-ty

kvakvs commented on TLM/TLM/Util/PriorityRoad.cs:
```diff
@@ -149,65 +291,74 @@ public static class PriorityRoad {
             ignoreLanes &= dir != ArrowDirection.Forward;
             ignoreLanes |= OptionsMassEditTab.PriorityRoad_AllowLeftTurns;
 
             Log._Debug($"ignorelanes={ignoreLanes} isSemiRabout={isSemiRabout}\n" +
```
Remove before merge. It pollutes the log with old irrelevant debug messages.

kvakvs commented on TLM/TLM/Util/PriorityRoad.cs:
```diff
@@ -441,5 +592,32 @@ public static class PriorityRoad {
             return forward + backward;
         }
+
+        public static void ClearNode(ushort nodeId) {
```
A quick oneline `///` doc string, what this does

kvakvs commented on TLM/TLM/Util/RoundaboutMassEdit.cs:
```diff
@@ -215,6 +221,30 @@ public class RoundaboutMassEdit {
             return ret;
         }
 
+        public static bool IsRabout(List<ushort> segList, bool semi = false) {
```
Please prefer full word `IsRoundabout`

kvakvs commented on TLM/TLM/Util/RoundaboutMassEdit.cs:
```diff
@@ -174,22 +174,28 @@ public class RoundaboutMassEdit {
+            return true;
+        }
+
+        public void FixRabout(List<ushort> segmentList) {
```
Please prefer full word `FixRoundabout`

kvakvs commented on TLM/TLM/Util/RoundaboutMassEdit.cs:
```diff
@@ -215,6 +221,30 @@ public class RoundaboutMassEdit {
+                    ushort headNodeID = netService.GetHeadNode(prevSegmentID);
+                    bool isRAbout = IsPartofRoundabout(nextSegmentID, prevSegmentID, headNodeID);
+                    if (!isRAbout) {
+                        //Log._Debug($"segments {prevSegmentID} and {nextSegmentID} with node:{headNodeID} are not part of a roundabout");
```
This log can be deleted before merge

kianzarrin replied:
is useful for future debugging though.

kvakvs commented on TLM/TLM/Util/RoundaboutMassEdit.cs:
```diff
@@ -320,5 +350,25 @@ public class RoundaboutMassEdit {
             return false;
         }
+
+        public void ClearNode(ushort nodeId) {
```
`/// <summary>Removes lane connectors from all segments connected to this node.</summary>`

kianzarrin replied:
it does much more than that. It clears all traffic rules related to roundabout. Note:
```
 PriorityRoad.ClearNode(nodeId);
```

kvakvs commented on TLM/TLM/Util/RoundaboutMassEdit.cs:
```diff
@@ -320,5 +350,25 @@ public class RoundaboutMassEdit {
+            });
+        }
+
+        public void ClearRabout(List<ushort> segmentList) {
```
Please prefer full name `ClearRoundabout`
/// <summary>Removes lane connectors from all nodes in this roundabout.</summary>

kvakvs commented on TLM/TLM/Util/PriorityRoad.cs:
```diff
@@ -9,18 +9,86 @@ namespace TrafficManager.Util {
+    using static TrafficManager.UI.SubTools.PrioritySignsTool;
+    using TrafficManager.API.Manager;
 
     public static class PriorityRoad {
```
Please add one line `///` summary for class.

kvakvs commented on TLM/TLM/Util/PriorityRoad.cs:
```diff
@@ -51,11 +119,51 @@ public static class PriorityRoad {
+        /// with <paramref name="otherSegmentId"/> .
+        /// </summary>
+        /// <param name="segmentId"></param>
+        /// <param name="otherSegmentId"></param>
```
Please don't leave these empty. Even one word is better, like `Segment 1.` Having these empty makes IDE sad.

kianzarrin commented on TLM/TLM/UI/RoadSelectionPanels.cs:
```diff
@@ -0,0 +1,388 @@
+        /// </summary>
+        /// <param name="reset">if true, deactivates all buttons</param>
+        public void Refresh(bool reset = false) {
+            Log._Debug($"Refresh called Function mode is {Function}\n");
```
refresh is called when user changes the road selection (or when he clicks on a button). This one - unlike the others - is useful debug info

kianzarrin replied:
come to think of it I do print debug info before calling Refresh so its fine.

kianzarrin commented on TLM/TLM/UI/RoadSelectionPanels.cs:
```diff
@@ -0,0 +1,388 @@
+                }
+                public int Width => 40;
+
+                public int Height => 40;
```
I was so much focused on get this freaking thing working without causing game crash that I forgot about Width and Height. It should be removed as part of merge I guess.

kvakvs commented on TLM/TLM/UI/RoadSelectionPanels.cs:
```diff
@@ -355,8 +359,21 @@ public abstract class ButtonExt : BaseMenuButton {
+                internal static float GetButtonSize() {
+                    var scaledSize
+                        = U.UIScaler.ScreenSizeSmallestFraction(0.0208f, 0.037f) *
+                          U.UIScaler.GetUIScale();
```
Suggested change (i will be applying similar change in my code too) makes it easier to see the origin of the fraction constants:
```
const float SUGGESTED_SIZE = 40f;
const float FRAC_X = SUGGESTED_SIZE / 1920f;
const float FRAC_Y = SUGGESTED_SIZE / 1080f;
...
U.UIScaler.ScreenSizeSmallestFraction(FRAC_X, FRAC_Y)
```

kianzarrin commented:
@kvakvs 
I reverted the scaling code because I think it was redundant in my case. As you can see from the images has the right size no matter the resolution. I tried with lowest resolution and 4K on my monitor and TV. More over using mod tools I can see the vanilla panel sizes that contain my UI are independent of resolution.

I tried to experiment with the scaling but it caused crashes because the sizes are unavailable on Awake. I tried to set them on Start() but that was too late and caused distortions. On the top of that I think I should not write code that does not do anything (as I said the sizes are correct no matter the resolution).

Should we postpone this matter of scaling to a later point?

kvakvs commented:
As you're embedding into existing UI, the scaling might be the wrong thing to do here.
As I noted before if you control entire window, then scaling is definitely the right thing to do.

kianzarrin commented:
>As you're embedding into existing UI, the scaling might be the wrong thing to do here.
As I noted before if you control entire window, then scaling is definitely the right thing to do.

I am embedding into existing UI.

kianzarrin commented:
I fixed all the problems by handling events with one frame delay. I also disabled Massedit overlay as I was unable to fix it see  #789
This is ready for review.
tagging @aubergine10 @kvakvs @krzychu124

kvakvs commented:
I took a closer look at the icons you used. Octagon looks distorted on the screenshot. Most buttons are using red palette while the only thing deserving to use red is the delete button.
Possible solution, make the stop octagon and yield sign less saturated red, or change silver or white color. Fill them with some darker shade of gray to give them some weight. The main road yellow square is only recognized in EU but at least consider adding a contrast black triangle between yellow and white fill.

kianzarrin commented:
@kvakvs  I know the Icons are bad. this is my first time drawing UI. Do you remember saying [this](https://github.com/CitiesSkylinesMods/TMPE/issues/542#issuecomment-577621846):
> Make a temporary ugly sprite. Show it to us and we might be able to improve it.
Or someone can feel inspiration and combine 1-2 free icons for you in Ms Paint.

> Octagon looks distorted on the screenshot.

Yeah some of the screen shots are super low quality. don't rely on them. I think the problem is I need to center the octagon.

> Most buttons are using red palette while the only thing deserving to use red is the delete button. 

Yield and stop signs are red in real life. why shouldn't they be red here?

>The main road yellow square (is only recognized in EU but at least) consider adding a contrast black triangle between yellow and white fill, or maybe instead use darker shade of orange.

I don't understand what do you mean by  triangle. what kind of triangle?

kvakvs commented:
Unclear behaviour.
I got all buttons grey upon game loading. After i dragged streets here and there in adjust mode, it eventually enabled buttons. There is some delay between i touch some road and buttons switch tom some appropriate state for that road. Selecting a roundabout street and clicking roundabout produced no visible effect. I.e. roundabout rules were not set.

kianzarrin commented:
> I got all buttons grey upon game loading. After i dragged streets here and there in adjust mode, it eventually enabled buttons. 

If you do not have any road selection those buttons obviously don't do anything.

> There is some delay between i touch some road and buttons switch tom some appropriate state 

Its delayed by one frame. Otherwise the game crashes. It took me days of debugging to realize that. Its a very short delay so I am not sure why it bothers you.

> Selecting a roundabout street and clicking roundabout produced no visible effect. I.e. roundabout rules were not set.

you need to select the whole roundabout. if the selection is not roundabout the roundabout button would be disabled.

Also remember I have disabled mass edit overlay due to #789 . But next round I will upload a partial workaround.

kianzarrin commented:
Thanks @kvakvs from pressuring me to fix the inconvenient UI.
I fixed several problems:
- Mass edit overlay now works at all times. (#789 fixed)
- I load advisor and guide locales when level is loaded. that means you don't see empty tutorial message. The tutorial message explains everything you asked me to (i didn't add anything to the explanation it was always there but you just couldn't see it before activating traffic manager tool at least once)

One problem remains though. if you select the road for the first time you need to press the green road button to be able to select the road (this is explained in the tutorial)

kianzarrin commented:
I removed the road selection panel from RoadWorldInfoPanel because RoadWorldInfoPanel does not actually change road selection. The base game is a confusing. Clicking on a road does not change road selection. enabling the priority toggle does not change last road selection but rather it changes rules for the currently selected road (it follows the name). Nevertheless Prioirt toggle is disabled if there is no road selection.  so its inconsistent.

I will now try to fix that inconsistency.

also loading a game while still in game causes null exceptions. I need to fix that to.

kianzarrin commented:
I fixed all the problems now:
- Mass edit overlay works from the start.
- translations are loaded from the start.
- no more disabled buttons when you are selecting a valid road.
- I added the buttons to RoadWorldInfoPanel again.

@kvakvs is that to your liking?

aubergine10 commented:
Looks like this branch is based on 11.1.0 tag - can you rebase to master?

aubergine10 commented:
Advisor tutorial: Good, made it clear how it works and what to expect. 👍 

Tooltips are good, explaining what the buttons do and when they will be disabled. 👍 

----

**Routes** tab:

There's no green "road route" bar on this view, and no road selection outlines, so it is not clear to user what bit(s) of road will be updated if they click the button.

![image](https://user-images.githubusercontent.com/1386719/78083177-a9e2be80-73ac-11ea-9bdb-97516ea0cfe3.png)

(We know how it all works, but user's don't).

Suggestion: When "Routes" tab is active, remove the overlays and buttons.

----

**Junctions** tab:

If I had clicked a road in either "Routes" or "Adjust Roads" tabs, the road info panel and overlays remain visible in the "Junctions" tab.

![image](https://user-images.githubusercontent.com/1386719/78083231-c54dc980-73ac-11ea-90d1-2aefcd8e367a.png)

This is super-confusing from user perspective as they now have both TMPE and vanilla junction stuff on screen at the same time, with the TMPE overlays obscuring the vanilla junction icons.

The other issue here is that it's not clear to user the difference between vanilla vs. TMPE features.

At some point, likely new PR, we should consider somehow replacing the vanilla icons with TMPE stuff (or find better way to merge them; eg. vanilla Stop signs and traffic lights are used by road asset authors to toggle meshes, props or decals on their road assets - so user still needs independent control over vanilla stop signs at least).

For now I suggest removing the bulk edit stuff while the "Junctions" tab is active.

----

**Adjust Roads** tab:

I found it a bit confusing that the UI is essenitally duplicated:

![image](https://user-images.githubusercontent.com/1386719/78082455-f75e2c00-73aa-11ea-9d4f-77434440bfd3.png)

The smaller panel that 'points to' the selected road feels much more natural place to be interacting with the tools.

Problem with that panel: If I close it, I can't just click the same route to show it again, so I have to click another route then re-select the original route I want to edit = cumbersome.

Maybe we could alter vanilla functionality so that when I click an already-selected route (but not dragging it's nodes) the panel reapepars? And then just have the bulk tools on that smaller panel - remove them from the bigger panel?

----

Positioning of the junction restriction icons in the overlay is still very grim from usability perspective (something for a separate PR, but just flagging it up again).

----

**Roundabouts:**

This confused me at first:

![image](https://user-images.githubusercontent.com/1386719/78083660-01355e80-73ae-11ea-91a4-18d6cb0d049d.png)

I then realised I'd made it with 2-way road. Maye add a note to toolip and/or advisor tutorial stating that roundabout must be one-way road?

Also, the stay in lane feature (at least with 2-lane roundabout) has an obvious issue here:

![image](https://user-images.githubusercontent.com/1386719/78083930-d0a1f480-73ae-11ea-9635-f22348b49b4b.png)

----

Those are the only usability/functionality issues I can find currently, I'll keep testing and report back if any others are found.

aubergine10 commented:
BTW, the buttons/icons are looking really good! 👍

kianzarrin commented:
> @aubergine10 **click arrow headings below to see feedback**

----

<details><summary>Routes tab:</summary>

> There's no green "road route" bar on this view, and no road selection outlines, so it is not clear to user what bit(s) of road will be updated if they click the button.

In the Vanilla version you can setup the priority road without seeing the green route. In the vanilla version usually  you can either see the green path or the 



![image](https://user-images.githubusercontent.com/1386719/78083177-a9e2be80-73ac-11ea-9bdb-97516ea0cfe3.png)

(We know how it all works, but user's don't).

Suggestion: When "Routes" tab is active, remove the overlays and buttons.</details>


----

<details><summary>Junctions tab:</summary>

If I had clicked a road in either "Routes" or "Adjust Roads" tabs, the road info panel and overlays remain visible in the "Junctions" tab.

![image](https://user-images.githubusercontent.com/1386719/78083231-c54dc980-73ac-11ea-90d1-2aefcd8e367a.png)

This is super-confusing from user perspective as they now have both TMPE and vanilla junction stuff on screen at the same time, with the TMPE overlays obscuring the vanilla junction icons.

The other issue here is that it's not clear to user the difference between vanilla vs. TMPE features.

At some point, likely new PR, we should consider somehow replacing the vanilla icons with TMPE stuff (or find better way to merge them; eg. vanilla Stop signs and traffic lights are used by road asset authors to toggle meshes, props or decals on their road assets - so user still needs independent control over vanilla stop signs at least).

For now I suggest removing the bulk edit stuff while the "Junctions" tab is active.</details>

----

<details><summary>Adjust Roads tab:</summary>

I found it a bit confusing that the UI is essenitally duplicated:

![image](https://user-images.githubusercontent.com/1386719/78082455-f75e2c00-73aa-11ea-9d4f-77434440bfd3.png)

The smaller panel that 'points to' the selected road feels much more natural place to be interacting with the tools.

Problem with that panel: If I close it, I can't just click the same route to show it again, so I have to click another route then re-select the original route I want to edit = cumbersome.

Maybe we could alter vanilla functionality so that when I click an already-selected route (but not dragging it's nodes) the panel re-appears? And then just have the bulk tools on that smaller panel - remove them from the bigger panel?</details>

----

Positioning of the junction restriction icons in the overlay is still very grim from usability perspective (something for a separate PR, but just flagging it up again).

----

<details><summary>Roundabouts:</summary>

This confused me at first:

![image](https://user-images.githubusercontent.com/1386719/78083660-01355e80-73ae-11ea-91a4-18d6cb0d049d.png)

I then realised I'd made it with 2-way road. Maye add a note to toolip and/or advisor tutorial stating that roundabout must be one-way road?

Also, the stay in lane feature (at least with 2-lane roundabout) has an obvious issue here:

![image](https://user-images.githubusercontent.com/1386719/78083930-d0a1f480-73ae-11ea-9635-f22348b49b4b.png)</details>

----

Those are the only usability/functionality issues I can find currently, I'll keep testing and report back if any others are found.

The routs tab and the Junctions tab are not meant to do anything WRT road selection. The only reason why you have mass edit overlay visible is because the `RoadWorldInfoView` panel is open.

If you click on a road while Routes t
In the vanilla game the priority toggle is when you click on a roads to see routes. Would it be a good idea to continue to show my road selection panel

aubergine10 commented:
@kianzarrin  did you mean to close this PR?

kianzarrin commented:
> @aubergine10 

**click arrow headings below to see feedback**

----

<details><summary>Routes tab:</summary>

> There's no green "road route" bar on this view, and no road selection outlines, so it is not clear to user what bit(s) of road will be updated if they click the button.

In the Vanilla version you can setup the priority road without seeing the green route. 

![Screenshot (754)](https://user-images.githubusercontent.com/26344691/78145453-93cf0f80-7439-11ea-9183-0489d87cbe32.png)

Its a bit problematic because you cannot see the stop signs unless you go to the junctions view. and if there are traffic lights stop signs are not setup and you would not notice unless if you go to junctions view. its a good thing I enabled mass edit overlay. 

![Screenshot (753)](https://user-images.githubusercontent.com/26344691/78145541-b103de00-7439-11ea-9872-94e6c1700383.png)

In the vanilla version (as you have noticed) usually  you can either see the green path or the `RoadWorldInfoPanel` panel. The objective of the author probabely was that the green route view is for adjusting the road but to setup prioirty road you don't need the green route. We don't have to do that of course we can change vanilla behaviour (but first we need to understand it).
That is a good idea. Another alternative is not to open the `RoadWorldInfoPanel` at all when clicking roads when Traffic Routes panel is open. your idea is much easier to implement though.

> Suggestion: When "Routes" tab is active, remove the overlays and buttons.

That is a good idea. Another alternative is not to open the `RoadWorldInfoPanel` at all when clicking roads when Traffic Routes panel is open. your idea is much easier to implement though.
</details>

----

<details><summary>Junctions tab:</summary>

> If I had clicked a road in either "Routes" or "Adjust Roads" tabs, the road info panel and overlays remain visible in the "Junctions" tab.

If you switch to any info view the `RoadWorldInfoPanel` closes down. I can do the same with junctions tabs. In other words when user clicks on junctions tab, the `RoadWorldInfoPanel` closes. Do you agree?


> At some point, likely new PR, we should consider somehow replacing the vanilla icons with TMPE stuff (or find better way to merge them; eg. vanilla Stop signs and traffic lights are used by road asset authors to toggle meshes, props or decals on their road assets - so user still needs independent control over vanilla stop signs at least).

maybe user can toggle yeild sign here as well :)

I have some plans for the junctions tab though I have not consolidated my thoughts yet. something like the road selection panel but for junctions. 
Just like road selection panel can be used for intent (#710) so can the junctions panel. we may forexampel add functionality for continues junction barrier.
</details>

----

<details><summary>Adjust Roads tab:</summary>

> The smaller panel that 'points to' the selected road feels much more natural place to be interacting with the tools.
Problem with that panel: If I close it, I can't just click the same route to show it again, so I have to click another route then re-select the original route I want to edit = cumbersome.

```c#
// this code is relevant.
	public void OnAdjustRoadButton()
	{
		Singleton<InfoManager>.instance.SetCurrentMode(InfoManager.InfoMode.TrafficRoutes, 2);
		Hide(); // this hides the RoadWorldInfoPanel when you switch to TrafficRoutes
	}
```

I think the autor didn't wanted these to panels to be open together.

> Maybe we could alter vanilla functionality so that when I click an already-selected route (but not dragging it's nodes) the panel re-appears? And then just have the bulk tools on that smaller panel - remove them from the bigger panel?

It is also possible to make `RoadWorldInfoPanel` not to open at all so that the buttons are in only one of the pannels.
</details>

----

> Positioning of the junction restriction icons in the overlay is still very grim from usability perspective (something for a separate PR, but just flagging it up again).

what if I only bring them up when user clicks one of my road selection buttons - or optionally when he is in road adjust panel ?

----

<details><summary>Roundabouts:</summary>


> I then realised I'd made it with 2-way road. Maye add a note to toolip and/or advisor tutorial stating that roundabout must be one-way road?

OK.

> Also, the stay in lane feature (at least with 2-lane roundabout) has an obvious issue here:
![image](https://user-images.githubusercontent.com/1386719/78083930-d0a1f480-73ae-11ea-9635-f22348b49b4b.png)

This is issue #617 and is already fixed in PR #706 (not merged yet).

</details>

----

So basically there are 2 possiblalities:
1- remove buttons from Road adjust panel and make it easier to open RoadWorldInfoPanel.
2- Keep buttons in both panels but make it impossible to open both at once. 

@aubergine10 please tell me which one?

kvakvs commented:
Remove the ambiguity, and then it looks like variant 1.

aubergine10 commented:
Still pondering this, I think there might be an option 3, just testing some stuff in game.

aubergine10 commented:
What about something like this...

Info views:

* Traffic Routes > **Routes** tab:
    * Remove the mass edit stuff
* Traffic Routes > **Junctions** tab:
    * Remove the mass edit stuff
* Traffic Routes > **Adjust Roads** tab:
   * Remove buttons from the Adjust Roads tab panel
   * Allow `RoadWorldInfoPanel` to be displayed when clicking already-selected green route

Normal map view:

* While TM:PE toolbar is visible (but no tool selected) allow user to click any road/rail (even if road names are hidden) to get it's `RoadWorldInfoPanel`?
* Also show the green route line (minus circles, and maybe semi-transparent) for selected road.

This means I wouldn't need to enter info views at all, it would mean I could have road names hidden yet still easily go and rename them or alter the road routes and apply mass edits.

aubergine10 commented:
> > If I had clicked a road in either "Routes" or "Adjust Roads" tabs, the road info panel and overlays remain visible in the "Junctions" tab.

> If you switch to any info view the RoadWorldInfoPanel closes down. I can do the same with junctions tabs. In other words when user clicks on junctions tab, the RoadWorldInfoPanel closes. Do you agree?

Agree - when switching tabs within Traffic Routes info view, the persistent overlay should be removed.

----

> > Positioning of the junction restriction icons in the overlay is still very grim from usability perspective (something for a separate PR, but just flagging it up again).

> what if I only bring them up when user clicks one of my road selection buttons - or optionally when he is in road adjust panel ?

It's not issue of when they are displayed, it's their position. See #633 - this is an issue in the persistent overlay for junction restrictions.

kianzarrin commented:
this review is from Nov-2009 so If you want me to add all these features you should also review them. Don't want this PR to become a relic of ancient past!

kianzarrin commented:
@aubergine10 In future we MIGHT code the Road Adjust tool such that when the user presses the high priority road button for example and then extends a road, traffic rules are applied to the new segments (#710). In that case does it make sense to keep the buttons in the Road Adjust tab? 

We could change the UI later but it might confusing if we change UI.

aubergine10 commented:
> this review is from Nov-2009 so If you want me to add all these features you should also review them. Don't want this PR to become a relic of ancient past!

What was that referring to? The junction restriction overlays? If so, that's definiately separate issue that can be done as speparate PR. If something else, let me know.

> EDIT: I move this to discord.

Will follow up via discord on the other stuff. Just need to run some errrands first so may be an hour or so.

kianzarrin commented:
@aubergine10 I uploaded the changes you requested. see commit message for details. Two things I did differently:
1- show Road Info panel when user clicks on road when Road adjust panel is open. (implementing this will make it impossible to adjust road)
2- Hide Road info panel when user modifies path. 

Instead I Show Road Info Panel automatically whenever user goes to road info panel. I will explain in a video why when i sort out some technical difficulties with my computer.

aubergine10 commented:
I love being able to select roads while TMPE toolbar is open!

----

<details><summary>Highlight whole route?</summary>
I selected this segment, clicked "Give way" button, and two junctions were updated:
![image](https://user-images.githubusercontent.com/1386719/78466865-5353e800-76fe-11ea-8530-e0868294e6a0.png)
Could we highlight all segments along the road?

Compare to how Shift+Click priority signs works - so basically this in green:
![image](https://user-images.githubusercontent.com/1386719/78466983-df1a4400-76ff-11ea-93bf-fc93e926852d.png)
</details>

----

<details><summary>Remove/replace routes arrow top-left of panel?</summary>
This is vanilla issue (really bad UI design) - the road world info panel has routes icon top left which is confusing (as over on the top-right there's the acutal routes button lol):
![image](https://user-images.githubusercontent.com/1386719/78466889-acbc1700-76fe-11ea-96ff-b92b8ddb7f00.png)
Could that be removed? Alternatively, replace with road icon?</details>

----

<details><summary>Is it supposed to select rails?</summary>
![image](https://user-images.githubusercontent.com/1386719/78466926-20f6ba80-76ff-11ea-84dc-1fde295dace2.png)
</details>

----

<details><summary>Right-click to close?</summary>
While TMPE toolbar open (no tool selected), if I select a road the world info panel pops up... but it's cumbersome clicking it's close button to remove it. Could we make it so a right-click deselects that road, closing the world info panel?</details>

----

<details><summary>Roundabout policies not working?</summary>
I have these two set:

![image](https://user-images.githubusercontent.com/1386719/78467138-75029e80-7701-11ea-82db-c69328e90d9c.png)

Dedicated exit lane doesn't seem to be applied (note: I'm using UK Roads which don't have arrow decals so I used lane arrow tool to illustrate what the lane arrows are):

![image](https://user-images.githubusercontent.com/1386719/78467123-513f5880-7701-11ea-9c94-6b245c46a01d.png)

And the roads entering roundabout (from bottom in this case) didn't have 'stay in lane' set:
![image](https://user-images.githubusercontent.com/1386719/78467172-b004d200-7701-11ea-8790-868357a2dd44.png)
</details>

kianzarrin commented:
@aubergine10 
> Could we highlight all segments along the road?

YES!

> Remove/replace routes arrow top-left of panel?

YES!

> Is it supposed to select rails?

Oops!

> Could we make it so a right-click deselects that road, closing the world info panel?

What? where should I right click to close road? anywhere? anywhere outside UI panels? any road? any segment with the same name? selected segment?

> Dedicated exit lane doesn't seem to be applied 

There is an exception to ignore lane arrow fixing when the road is not going forward. It is because the math is too hard and can lead to too many combinations. In any case it is not part of this review. it should be part of priority road/roundabout phase 2 I guess?

> And the roads entering roundabout (from bottom in this case) didn't have 'stay in lane' set:

The tooltip says "Vehicles shall not switch lanes TOO CLOSE to the roundabout". Only when the node is too close to the roundabout do we ask cars to stay in lane outside of roundabout.

kianzarrin commented:
@aubergine10 I fixed some of the issues you raised :
- right-click closes road selection panel.
- train cannot be selected.
- hovering over road selection buttons shows the NetAdjust path (it used to when TMPE was disabled but now it does it even when TMPE is enabled).

krzychu124 commented on TLM/TLM/Util/RoadSelectionUtil.cs:
```diff
@@ -0,0 +1,173 @@
+        }
+
+        public static void Release() {
+            Instance.OnChanged = null;
```
NullpointerException possible

krzychu124 commented on TLM/TLM/Util/RoadSelectionUtil.cs:
```diff
@@ -0,0 +1,173 @@
+            IsNetAdjustMode(InfoManager.instance.CurrentMode, (int)InfoManager.instance.CurrentSubMode);
+
+        internal static bool IsNetAdjustMode(InfoManager.InfoMode mode, int subModeIndex) =>
+            (mode == InfoManager.InfoMode.TrafficRoutes) &
```
looks like a typo, shouldn't be `&&` here?

krzychu124 commented on TLM/TLM/Util/RoadSelectionUtil.cs:
```diff
@@ -0,0 +1,173 @@
+                    List<ushort> ret = new List<ushort>();
+                    for (int i = 0; i < Length; ++i) {
+                        ushort segmentId = path.m_buffer[i];
+                        ret.Add(segmentId);
```
you could inline that ;)

krzychu124 commented on TLM/TLM/Util/RoadSelectionUtil.cs:
```diff
@@ -0,0 +1,173 @@
+        public List<ushort> Selection {
+            get {
+                if (Length > 0) {
+                    FastList<ushort> path = GetPath();
```
Isn't it null possible here? If not why do you use `GetPath()?` in `Length` property?

krzychu124 commented on TLM/TLM/Util/RoadSelectionUtil.cs:
```diff
@@ -0,0 +1,173 @@
+
+            public override void OnUpdate(float realTimeDelta, float simulationTimeDelta) {
+                try {
+                    if (Instance == null) {
```
Instance of what? That why I hate nested classes...

krzychu124 commented on TLM/TLM/UI/TrafficManagerTool.cs:
```diff
@@ -442,13 +511,41 @@ public class TrafficManagerTool
+                            SegmentStopCriterion.None,
+                            (_) => true);
+                    }
+                    RoadSelectionUtil.SetRoad(HoveredSegmentId, segmentList);
```
Nullpointer possible, `segmentList` can be null, because of changes to `Segment.Traverser()` in this PR

kianzarrin commented:
@krzychu124 @aubergine10 I made/responded to all the changes you requested. Ready for the next round!

kianzarrin commented:
Lets get this over with before more conflicts arise!

kianzarrin commented:
@aubergine10 Although the code functionally works i was unable to hide the RoadWorldInfoPanel without null reference exception in output log. So I commented out the code and created #822
I still am able to open it though.

@krzychu124 I particularly want to draw your attention to multi-threading. I created several methods in TrafficManagerTool.cs to handle/expose RenderOverlay and OnToolGUI without causing infinite recursion. Do you like my design?

aubergine10 commented:
This is freaking awesome!!!!

![image](https://user-images.githubusercontent.com/1386719/79054699-f9dc4380-7c3e-11ea-833a-3c87c53dbc80.png)

kianzarrin commented:
> I'll review the code when you fix conflicts, because I have no idea which code you've merged and which you'd changed since last review.

There wasn't conflicts when I asked you to review it. They just keep pop-ing up!

> I have no idea what you do there, but that process is horribly slow (adds more than 80ms per frame) :( 

I'll try to compare the FPS difference between when I set PathVisible true or false. I also compare the FPS with Vanilla road selection tool where the path is rendered. If the speed is slow even in the vanila case then, it's not my fault. I probably can fix it by caching the path but its out of the scope of this review. Actually if even the vanilla game is slow, it might be more relevant to your FPS booster mod?

> Clicking on option like clear, hangs game for more than second (I can live with it).

Only if there are several junctions too fix. am I wrong?

> I don't know why you do whole highlighting stuff when button is disabled.

Oops my bad!

> ![image](https://user-images.githubusercontent.com/19638970/79056703-58162000-7c59-11ea-987a-cef9696ae3d7.png)
It doesn't look like roundabout 😄 

if it is one-way I assume semi-roundabout. Road selection is supposed to be the more flexible alternative to Priority tool for magic/complex roundabouts. The more restrictions I add to what counts as a roundabout, the less flexible the tool will become and the more likely it might exclude rare complex roundabouts. So its a compromise.

kianzarrin commented:
> I have no idea what you do there, but that process is horribly slow

Nice catch. I should have realized it by looking at the log file. There were two bugs that together caused the slowdown!

kianzarrin commented:
-  fixed slow render overlay.
- fixed showing path when button is disabled.
- fixed pressing control now activates mass edit overlay (when no tool is selected).

krzychu124 commented:
- fixed slow render overlay. ✔️ 
- fixed showing path when button is disabled. ✔️ 
- fixed pressing control now activates mass edit overlay (when no tool is selected)- _not sure that one, because Mass Edit (green segment selector) is enabled all the time when menu is visible_

I hope that can be explained, because I have no idea why lane arrows were changed (all configurable options were disabled in options)
[![](https://img.youtube.com/vi/b2X4ksPTTjI/0.jpg)](https://youtu.be/b2X4ksPTTjI)

kianzarrin commented:
- @aubergine10 thanks to @krzychu124 I manged to fixed #882 and now I can hide/show the `RoadWorldInfoPanel` at appropriate times. switching to junctions tab hides the panel. so does left clicking if TMPE tool is enabled. 
- @aubergine10 Can you please check and see how the road names work. has anything changed since sunset harbor? can you extend a road by changing road names?

kianzarrin commented:
> fixed pressing control now activates mass edit overlay (when no tool is selected)- not sure that one, because Mass Edit (green segment selector) is enabled all the time when menu is visible

kianzarrin commented:
> @krzychu124 fixed pressing control now activates mass edit overlay (when no tool is selected)- not sure that one, because Mass Edit (green segment selector) is enabled all the time when menu is visible

press the traffic manager button but do not select any tool. press control to see the mass edit overlay.

kianzarrin commented:
> While menu is visible I'm noticing significant fps drop ~15-20ms/frame in my case, so from 30fps it drops to ~20fps
only TM:PE subscribed and enabled

I cannot reproduce this problem:
![Screenshot (840)](https://user-images.githubusercontent.com/26344691/79203515-d59d8400-7e43-11ea-8bb2-08863e5d8274.png)
![Screenshot (838)](https://user-images.githubusercontent.com/26344691/79203528-d7674780-7e43-11ea-81ce-210eac38c092.png)
In any case opening the `RoadWorldInfoPanel` (or what you referred to as "menu") results in additional calls to TMPE render overlay and OnToolGUI in order to render persistent overlay. maybe you have a lot of traffic rules and showing the overlay takes a while? 

Can you please produce minimum test case so I can reproduce this problem:
- can you reproduce this problem in a brand new map with no traffic rules?
- disable all traffic rules in your city. Compare the speed difference between when  `RoadWorldInfoPanel` is closed and opened.

Thanks.

kianzarrin commented:
> @krzychu124  I hope that can be explained, because I have no idea why lane arrows were changed (all configurable options were disabled in options)
[![](https://img.youtube.com/vi/b2X4ksPTTjI/0.jpg)](https://youtu.be/b2X4ksPTTjI)

This is a problem that also exists in master (Don't know when it got in).  But I fixed it :)

This is the second time you detect a problem the fix to which is changing `&&` to `||`! :D

kianzarrin commented:
I fixed a couple of problems please see commit message.

krzychu124 commented on TLM/TLM/Util/RoundaboutMassEdit.cs:
```diff
@@ -320,5 +350,33 @@ public class RoundaboutMassEdit {
+        /// Clears all rules put by RoundAboutMassEdit.FixRoundabout()
+        /// </summary>
+        /// <param name="segmentList"></param>
+        public void ClearRoundabout(List<ushort> segmentList) {
```
Doesn't that method clear the same node multiple times?

krzychu124 replied:
@kianzarrin no comment?

krzychu124 commented on TLM/TLM/Util/RoundaboutMassEdit.cs:
```diff
@@ -223,15 +253,15 @@ public class RoundaboutMassEdit {
-                    isRAbout = true;
+                    isRoundabout = true;
                 } else if (Contains(nextSegmentId)) {
-                    isRAbout = false;
+                    isRoundabout = false;
```
It looks like a bug. `isRoundabout` cannot be `true` here so there is no point to set it to false again

krzychu124 replied:
@kianzarrin I think you didn't understand my point here. 
I meant that check if nextSegmentId is inside list is completely pointless because you don't do anything regardless of result. That's why is think it is a bug here, because why bother checking for specific case when no different action is performed if occurs?

krzychu124 commented on TLM/TLM/Util/RoundaboutMassEdit.cs:
```diff
@@ -173,23 +173,29 @@ public class RoundaboutMassEdit {
+        }
+
+        public void FixRoundabout(List<ushort> segmentList) {
+            this.segmentList = segmentList;
+            int count = segmentList?.Count ?? 0;
```
Do you really want `NullReferenceException` here? 
Pretty unfortunate name for parameter, btw

krzychu124 replied:
@kianzarrin it's still not resolved!
Case is simple. Why you are afraid of `null` in `segmentList`, but the same null you are setting to `this.segmentList` field and inside other methods you use `this.segmentList.Clear()` without bothering with null check?

I don't understand why did you overwritten field value but still use param for next operations. I think that method does to much things at once

kianzarrin commented:
>> @krzychu124  Edit:
 While menu is visible I'm noticing significant fps drop ~15-20ms/frame in my case, so from 30fps it drops to ~20fps
 _only TM:PE subscribed and enabled_

> I cannot reproduce this problem

You didn't answer this. I take it that this is OK by now?

kianzarrin commented:
@krzychu124 I made a few lines of simple changes. Can you please approve this.

aubergine10 commented:
Regarding the performance issue while tmpe toolbar is open, like Kian I also don't see any drop in frame rate. Could be hardware specific though.

krzychu124 commented:
> Regarding the performance issue while tmpe toolbar is open, like Kian I also don't see any drop in frame rate. Could be hardware specific though.

Fps drop was caused by excessive log spam.

krzychu124 commented on TLM/TLM/UI/TrafficManagerTool.cs:
```diff
@@ -348,6 +365,58 @@ public class TrafficManagerTool
+        {
+            MassEditOVerlay.Show = ControlIsPressed || RoadSelectionPanels.Root.ShouldShowMassEditOverlay();
+            Log._Debug("RoadSelectionPanels.Root.RoadWorldInfoPanelExt.isActiveAndEnabled=" + RoadSelectionPanels.Root.RoadWorldInfoPanelExt.isActiveAndEnabled);
+            if (NetManager.instance.NetAdjust.PathVisible =
+                RoadSelectionPanels.Root.ShouldPathBeVisible()) {
```
How you fixed this?? Not even single comment about that.

kianzarrin commented:
> but I see possible bugs/typos

You are absolutely right!

> The following questions are still without answers nor even comment (yes/no)

@krzychu124  this is github being stupid. I did put comments but it does not show up in the conversation tab. you should go to the files changed tab to see the comments I put in the files changed tab. Can you please do that and respond to me?

kianzarrin commented:
@krzychu124 one more thing. Sometimes you have to press `Load Diff` for big files otherwise you might not see the comments.

aubergine10 commented:
![image](https://user-images.githubusercontent.com/1386719/79618433-f266df80-8101-11ea-8196-3894676243ab.png)
